frontend/apps/reader/modules/readerrolling.lua

```diff
@@ -1086,4 +1107,212 @@ function ReaderRolling:showEngineProgress(percent)
+        -- (see ReaderTypeset:onReadSettings() for the logic of block_rendering_mode)
+        local g_block_rendering_mode = G_reader_settings:readSetting("copt_block_rendering_mode")
+        if g_block_rendering_mode ~= 0 then -- default is not "legacy"
+            if not g_block_rendering_mode then -- nil means: use default
+                g_block_rendering_mode = 3 -- default in ReaderTypeset:onReadSettings()
+            end
+            -- This setting is actually saved by self.ui.document.configurable
+            local block_rendering_mode = self.ui.document.configurable.block_rendering_mode
+            if block_rendering_mode == 0 then
+                self.ui.document.configurable.block_rendering_mode = g_block_rendering_mode
+                logger.info("  block_rendering_mode switched to", g_block_rendering_mode)
+            end
+        end
+
+        -- No need for "if doc:hasCacheFile() then doc:invalidateCacheFile()", as
+        -- a change in gDOMVersionRequested has crengine trash previous cache file.
+    end
+
+    -- Check all xpointers
+    applyFuncToXPointersSlots(checkAndCount)
+    logger.info(T("%1 xpointers checked: %2 found (%3 changed) - %4 lost",
+                nb_xpointers, nb_xpointers_found, nb_xpointers_changed, nb_xpointers_lost))
+    if nb_xpointers_lost > 0 then
+        logger.warn("Lost xpointers:")
+        for k, v in pairs(lost_xpointer_info) do
+            logger.warn("  ", k, ":", v)
+        end
+    end
+
+    local text = _([[
+This book has been first opened, and handled since, by some old version of the rendering code.
+It can now be migrated to be handled by the newest and greatest rendering code.
+
+%1
+
+Do you wish to proceed with this upgrade, and reload your book with the newest code?]])
+    local details = {}
+    if nb_xpointers_lost == 0 then
+        table.insert(details, _([[All your bookmarks and highlights are valid and will be available after the migration.]]))
+    else
+        table.insert(details, T(_([[
+Note that %1 (out of %2) xpaths from your bookmarks and highlights aren't currently found in the book, and may have been lost. You might want to toggle Rendering mode between 'legacy' and 'flat', and re-open this book, and see if they are found again, before proceeding.]]),
+            nb_xpointers_lost, nb_xpointers))
+    end
+    if nb_xpointers_changed > 0 then
+        table.insert(details, T(_([[
+Note that %1 (out of %2) xpaths from your bookmarks and highlights have been normalized, and may not work on previous KOReader versions (if you're synchronizing your reading between multiple devices, you'll need to update KOReader on all of them).]]),
+            nb_xpointers_changed, nb_xpointers))
+    end
+    text = T(text, table.concat(details, "\n\n"))
+
+    UIManager:show(ConfirmBox:new{
+        text = text,
+        other_buttons = {{
+            {
+                -- this is the real cancel/do nothing
+                text = _("Ask me next time"),
```

Comment:
Bit long?

* Ask again
* Not now

Reply:
Not yet

Reply:
Ask later?

Reply:
The awful Apple way to deal with that in notifications is `Snooze`.

I kinda like `Not now`, FWIW ;).